pcap-linux.c

```diff
@@ -1456,90 +1441,37 @@ pcap_activate_linux(pcap_t *handle)
 	 * Otherwise, open a SOCK_RAW.
 	 */
 	ret = activate_sock(handle, is_any_device);
-	if (ret < 0) {
+	if (ret < 0)
 		/*
 		 * Fatal error; the return value is the error code,
 		 * and handle->errbuf has been set to an appropriate
 		 * error message.
 		 */
 		return ret;
-	} else {
-#ifdef HAVE_PACKET_RING
-		/*
-		 * Success.
-		 * Try to use memory-mapped access.
-		 */
-		switch (activate_mmap(handle, &status)) {
-
-		case 1:
-			/*
-			 * We succeeded.  status has been
-			 * set to the status to return,
-			 * which might be 0, or might be
-			 * a PCAP_WARNING_ value.
-			 *
-			 * Set the timeout to use in poll() before
-			 * returning.
-			 */
-			set_poll_timeout(handlep);
-			return status;
-
-		case 0:
-			/*
-			 * Kernel doesn't support it - just continue
-			 * with non-memory-mapped access.
-			 */
-			break;
-
-		case -1:
-			/*
-			 * We failed to set up to use it, or the
-			 * kernel supports it, but we failed to
-			 * enable it.  status has been set to the
-			 * error status to return and, if it's
-			 * PCAP_ERROR, handle->errbuf contains
-			 * the error message.
-			 */
-			goto fail;
-		}
-#endif /* HAVE_PACKET_RING */
-	}
 
 	/*
-	 * We set up the socket, but not with memory-mapped access.
+	 * Success. Set up memory mapping.
 	 */
-	if (handle->opt.buffer_size != 0) {
+	ret = activate_mmap(handle, &status);
+	if (ret == 1) {
 		/*
-		 * Set the socket buffer size to the specified value.
+		 * We succeeded.  status has been
+		 * set to the status to return,
+		 * which might be 0, or might be
+		 * a PCAP_WARNING_ value.
+		 *
+		 * Set the timeout to use in poll() before
+		 * returning.
 		 */
-		if (setsockopt(handle->fd, SOL_SOCKET, SO_RCVBUF,
-		    &handle->opt.buffer_size,
-		    sizeof(handle->opt.buffer_size)) == -1) {
-			pcap_fmt_errmsg_for_errno(handle->errbuf,
-			    PCAP_ERRBUF_SIZE, errno, "SO_RCVBUF");
-			status = PCAP_ERROR;
-			goto fail;
-		}
-	}
-
-	/* Allocate the buffer */
-
-	handle->buffer	 = malloc(handle->bufsize + handle->offset);
-	if (!handle->buffer) {
-		pcap_fmt_errmsg_for_errno(handle->errbuf, PCAP_ERRBUF_SIZE,
-		    errno, "malloc");
-		status = PCAP_ERROR;
-		goto fail;
+		set_poll_timeout(handlep);
+		return status;
 	}
 
 	/*
-	 * "handle->fd" is a socket, so "select()" and "poll()"
-	 * should work on it.
+	 * We failed to set up to use it.
+	 * status has been set to the error status to return and, if
+	 * it's PCAP_ERROR, handle->errbuf contains the error message.
 	 */
-	handle->selectable_fd = handle->fd;
```

Comment:
So where does this now set `handle->selectable_fd`?  It has to get set, otherwise applications won't be able to use `select()`/`poll()`/etc. on the socket.  (No, you're not supposed to use `pcap_fileno()` to get the selectable descriptor.)

Reply:
That was unintented. I'll fix that and take a closer look just in case. I'll try to be more careful next time. Thanks for your patience.